### PR TITLE
[Merged by Bors] - feat: port Data.Matrix.DMatrix

### PR DESCRIPTION
**Pull request by casavaca**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by casavaca** [comment]
Some Lint errors I don't know how to fix...

```
/- The `docBlame` linter reports:
DEFINITIONS ARE MISSING DOCUMENTATION STRINGS: -/
-- Mathlib.Data.Matrix.DMatrix
#check DMatrix.term_ᵀ /- definition missing documentation string -/

/- The `unusedArguments` linter reports:
UNUSED ARGUMENTS. -/
-- Mathlib.Data.Matrix.DMatrix
#check DMatrix.{u, u', v} /- argument 3 inst✝¹ : Fintype m, argument 4 inst✝ : Fintype n -/
```

File: Mathlib/Data/Matrix/DMatrix.lean [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  rfl
+#align dmatrix.map_apply DMatrix.map_apply
+
+@[simp]
+theorem map_map {M : DMatrix m n α} {β : m → n → Type w} {γ : m → n → Type z}
+    {f : ∀ ⦃i j⦄, α i j → β i j} {g : ∀ ⦃i j⦄, β i j → γ i j} :
+    (M.map f).map g = M.map fun i j x => g (f x) :=
+  by
+  ext
+  simp
```
Comment (dupuisf) [inline Mathlib/Data/Matrix/DMatrix.lean:68]:
```suggestion
    (M.map f).map g = M.map fun i j x => g (f x) := by ext; simp
```

Comment (dupuisf) [reply Mathlib/Data/Matrix/DMatrix.lean:68]: There are a few other instances of this in the file as well.

File: Mathlib/Data/Matrix/DMatrix.lean [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+def AddMonoidHom.mapDMatrix [∀ i j, AddMonoid (α i j)] {β : m → n → Type w}
+    [∀ i j, AddMonoid (β i j)] (f : ∀ ⦃i j⦄, α i j →+ β i j) : DMatrix m n α →+ DMatrix m n β
+    where
+  toFun M := M.map fun i j => @f i j
+  map_zero' := by simp
+  map_add' := DMatrix.map_add f
+#align add_monoid_hom.map_dmatrix AddMonoidHom.mapDMatrix
+
+@[simp]
+theorem AddMonoidHom.map_DMatrix_apply [∀ i j, AddMonoid (α i j)] {β : m → n → Type w}
```
Comment (dupuisf) [inline Mathlib/Data/Matrix/DMatrix.lean:200]:
```suggestion
theorem AddMonoidHom.mapDMatrix_apply [∀ i j, AddMonoid (α i j)] {β : m → n → Type w}
```

**Review by dupuisf — COMMENTED** [review]
Just a few cosmetic things, otherwise LGTM!

bors d+

**Pull request closed without merging**